### PR TITLE
Support for `NULL` tensors.

### LINKED ISSUE #837: Fail Gracefully with NULL tensors
NULLs passed to torch_tensor fail gracefully.
``` r
torch::torch_tensor(NULL)
#> Error in torch_tensor_cpp(data, dtype, device, requires_grad, pin_memory): R type not handled
```

NULLs passed to torch_tensor as integers "work".
```r
torch::torch_tensor(as.integer(NULL))
#> torch_tensor
#> [ CPULongType{0} ]
```

But if you subset that tensor, it causes R to crash.
```r
# Crashes Rstudio.
torch::torch_tensor(as.integer(NULL))[1]
```

This can happen relatively easily with a malformed dataset (for example, if you try to construct labels in a test set), and the crash is mysterious.

Should `torch_tensor` validate the `data` argument to give a friendlier error message for (any type of) NULL?

### PR DESCRIPTION
**Pull request by dfalbel**
Fix #837

### CODE REVIEW

**Comment by jonthegeek** [comment]
This crashes my Rstudio session (eventually; it hangs for a long time first):
```
torch::torch_tensor(integer(0))[1]
```

Likewise in the R console (hangs but never actually crashes, it just goes into "Not responding" state), and in the terminal (likewise just stops responding, never actually crashed; I had to abort both manually).

I tried to generate this as a nice reprex, but that just spins until I kill it, too.

```
> sessionInfo()
R version 4.2.0 (2022-04-22 ucrt)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 22000)

Matrix products: default

locale:
[1] LC_COLLATE=English_United States.utf8  LC_CTYPE=English_United States.utf8   
[3] LC_MONETARY=English_United States.utf8 LC_NUMERIC=C                          
[5] LC_TIME=English_United States.utf8    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.8.3         wordpiece.data_2.0.0 pillar_1.7.0         compiler_4.2.0      
 [5] tools_4.2.0          testthat_3.1.4       zeallot_0.1.0        digest_0.6.29       
 [9] bit_4.0.4            evaluate_0.15        jsonlite_1.8.0       rsample_0.1.1       
[13] lifecycle_1.0.1      tibble_3.1.7         pkgconfig_2.0.3      rlang_1.0.2         
[17] fastmatch_1.1-3      DBI_1.1.2            cli_3.3.0            yaml_2.3.5          
[21] parallel_4.2.0       xfun_0.31            coro_1.0.2           fastmap_1.1.0       
[25] knitr_1.39           furrr_0.3.0          dplyr_1.0.9          fs_1.5.2            
[29] hms_1.1.1            generics_0.1.2       vctrs_0.4.1          globals_0.15.0      
[33] bit64_4.0.5          prompt_1.0.1         tidyselect_1.1.2     glue_1.6.2          
[37] listenv_0.8.0        R6_2.5.1             processx_3.5.3       fansi_1.0.3         
[41] parallelly_1.31.1    luz_0.2.0            rmarkdown_2.14       tzdb_0.3.0          
[45] readr_2.1.2          purrr_0.3.4          tidyr_1.2.0          callr_3.7.0         
[49] dlr_1.0.1            magrittr_2.0.3       htmltools_0.5.2      codetools_0.2-18    
[53] ps_1.7.0             ellipsis_0.3.2       assertthat_0.2.1     future_1.26.1       
[57] torch_0.7.2.9000     utf8_1.2.2           crayon_1.5.1         brio_1.1.3    
```

**Comment by dfalbel** [comment]
Ok, I'll try to reproduce it later. Might be a Windows only problem. On MacOs I see:

``` r
torch::torch_tensor(integer(0))[1]
#> Error in Tensor_slice(tensor$ptr, environment(), drop = drop, mask = .d): index 0 is out of bounds for dimension 0 with size 0
#> Exception raised from applySelect at /Users/distiller/project/pytorch/aten/src/ATen/TensorIndexing.h:230 (most recent call first):
#> frame #0: c10::Error::Error(c10::SourceLocation, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >) + 98 (0x112310f12 in libc10.dylib)
#> frame #1: at::indexing::impl::applySelect(at::Tensor const&, long long, long long, long long, c10::Device const&, c10::ArrayRef<long long> const&) + 251 (0x131a9a5eb in libtorch_cpu.dylib)
#> frame #2: at::Tensor::index(c10::ArrayRef<at::indexing::TensorIndex>) const + 582 (0x131a98f26 in libtorch_cpu.dylib)
#> frame #3: _lantern_Tensor_index + 319 (0x116d015ff in liblantern.dylib)
#> frame #4: Tensor_slice(XPtrTorchTensor, Rcpp::Environment_Impl<Rcpp::PreserveStorage>, bool, Rcpp::Vector<19, Rcpp::PreserveStorage>) + 571 (0x115fb84cb in torchpkg.so)
#> frame #5: _torch_Tensor_slice + 310 (0x115e58976 in torchpkg.so)
#> frame #6: R_doDotCall + 1483 (0x10f897deb in libR.dylib)
#> frame #7: do_dotcall + 334 (0x10f89975e in libR.dylib)
#> frame #8: bcEval + 28581 (0x10f8d0795 in libR.dylib)
#> frame #9: Rf_eval + 385 (0x10f8c9141 in libR.dylib)
#> frame #10: R_execClosure + 2169 (0x10f8e8fb9 in libR.dylib)
#> frame #11: Rf_applyClosure + 471 (0x10f8e7da7 in libR.dylib)
#> frame #12: bcEval + 26782 (0x10f8d008e in libR.dylib)
#> frame #13: Rf_eval + 385 (0x10f8c9141 in libR.dylib)
#> frame #14: forcePromise + 172 (0x10f8e74bc in libR.dylib)
#> frame #15: Rf_eval + 522 (0x10f8c91ca in libR.dylib)
#> frame #16: forcePromise + 172 (0x10f8e74bc in libR.dylib)
#> frame #17: Rf_eval + 1124 (0x10f8c9424 in libR.dylib)
#> frame #18: Rf_evalList + 168 (0x10f8e75c8 in libR.dylib)
#> frame #19: Rf_eval + 1210 (0x10f8c947a in libR.dylib)
#> frame #20: Rf_evalList + 168 (0x10f8e75c8 in libR.dylib)
#> frame #21: Rf_eval + 1210 (0x10f8c947a in libR.dylib)
#> frame #22: do_if + 60 (0x10f8ea15c in libR.dylib)
#> frame #23: Rf_eval + 887 (0x10f8c9337 in libR.dylib)
#> frame #24: do_begin + 462 (0x10f8eb95e in libR.dylib)
#> frame #25: Rf_eval + 887 (0x10f8c9337 in libR.dylib)
#> frame #26: R_execClosure + 2169 (0x10f8e8fb9 in libR.dylib)
#> frame #27: Rf_applyClosure + 471 (0x10f8e7da7 in libR.dylib)
#> frame #28: bcEval + 26782 (0x10f8d008e in libR.dylib)
#> frame #29: Rf_eval + 385 (0x10f8c9141 in libR.dylib)
#> frame #30: R_execClosure + 2169 (0x10f8e8fb9 in libR.dylib)
#> frame #31: Rf_applyClosure + 471 (0x10f8e7da7 in libR.dylib)
#> frame #32: bcEval + 26782 (0x10f8d008e in libR.dylib)
#> frame #33: Rf_eval + 385 (0x10f8c9141 in libR.dylib)
#> frame #34: R_execClosure + 2169 (0x10f8e8fb9 in libR.dylib)
#> frame #35: Rf_applyClosure + 471 (0x10f8e7da7 in libR.dylib)
#> frame #36: bcEval + 26782 (0x10f8d008e in libR.dylib)
#> frame #37: Rf_eval + 385 (0x10f8c9141 in libR.dylib)
#> frame #38: R_execClosure + 2169 (0x10f8e8fb9 in libR.dylib)
#> frame #39: Rf_applyClosure + 471 (0x10f8e7da7 in libR.dylib)
#> frame #40: dispatchMethod + 775 (0x10f932fb7 in libR.dylib)
#> frame #41: Rf_usemethod + 735 (0x10f932c7f in libR.dylib)
#> frame #42: Rf_DispatchOrEval + 1285 (0x10f8edfe5 in libR.dylib)
#> frame #43: R_DispatchOrEvalSP + 429 (0x10f9ac24d in libR.dylib)
#> frame #44: do_subset + 31 (0x10f9ac07f in libR.dylib)
#> frame #45: Rf_eval + 887 (0x10f8c9337 in libR.dylib)
#> frame #46: do_eval + 625 (0x10f8ecff1 in libR.dylib)
#> frame #47: bcEval + 28581 (0x10f8d0795 in libR.dylib)
#> frame #48: Rf_eval + 385 (0x10f8c9141 in libR.dylib)
#> frame #49: R_execClosure + 2169 (0x10f8e8fb9 in libR.dylib)
#> frame #50: Rf_applyClosure + 471 (0x10f8e7da7 in libR.dylib)
#> frame #51: bcEval + 26782 (0x10f8d008e in libR.dylib)
#> frame #52: Rf_eval + 385 (0x10f8c9141 in libR.dylib)
#> frame #53: forcePromise + 172 (0x10f8e74bc in libR.dylib)
#> frame #54: Rf_eval + 1124 (0x10f8c9424 in libR.dylib)
#> frame #55: do_withVisible + 57 (0x10f8ed669 in libR.dylib)
#> frame #56: do_internal + 362 (0x10f930d6a in libR.dylib)
#> frame #57: bcEval + 29053 (0x10f8d096d in libR.dylib)
#> frame #58: Rf_eval + 385 (0x10f8c9141 in libR.dylib)
#> frame #59: R_execClosure + 2169 (0x10f8e8fb9 in libR.dylib)
#> frame #60: Rf_applyClosure + 471 (0x10f8e7da7 in libR.dylib)
#> frame #61: bcEval + 26782 (0x10f8d008e in libR.dylib)
#> frame #62: Rf_eval + 385 (0x10f8c9141 in libR.dylib)
#> frame #63: forcePromise + 172 (0x10f8e74bc in libR.dylib)
```

<sup>Created on 2022-05-31 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by jonthegeek** [comment]
I still haven't gotten it to build properly completely locally; I have to remotes::install_github() to install dev versions. Maybe it's something related to that?

Either way... probably make it fail more gracefully, and then it should be solved?

**Comment by dfalbel** [comment]
Ok! I was able to reproduce using R 4.2 instead of 4.1 and CRAN R. Trying now with dev version.
